src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1030,11 +1037,18 @@ async ValueTask DisposeAsyncCore()
             {
                 await Close(connectionClosing: false, async: true, isDisposing: true);
             }
-            catch (Exception e)
+            catch (Exception ex)
             {
-                _commandLogger.LogError(e, "Exception caught while disposing a reader", Connector.Id);
-                if (e is not PostgresException)
+                // In the case of a PostgresException (or multiple ones, if we have error barriers), the reader's state has already been set
+                // to Disposed in Close above; otherwise, we need to set it here.
```

Comment:
@vonzshik I see you added the condition to only set State to Disposed if the exceptions is not a PostgresException. I don't remember the context for that issue, but would it be problematic if we just always set it (either for any exception, or just in the finally)?

Reply:
One potential issue here is multiplexing: after a reader is disposed, we return it to the connection, but it happens before we set the `Disposed` state here.

Reply:
Thanks, I remember now. Updated the comment to make this clear.

Reply:
Well, there is also `CommandBehavior.CloseConnection`, but the idea is the same.